<!DOCTYPE HTML>
<!--
	Alpha by HTML5 UP
	html5up.net | @ajlkn
	Free for personal and commercial use under the CCA 3.0 license (html5up.net/license)
-->
<html>
	<head>
		<title>Eagle FAQ</title>
		<meta charset="utf-8" />
		<meta name="viewport" content="width=device-width, initial-scale=1" />
		<!--[if lte IE 8]><script src="assets/js/ie/html5shiv.js"></script><![endif]-->
		<link rel="stylesheet" href="assets/css/main.css" />
		<!--[if lte IE 8]><link rel="stylesheet" href="assets/css/ie8.css" /><![endif]-->
	</head>
	<body>
		<div id="page-wrapper">

			
				<header id="header">
					<nav id="nav">
						<ul>
							<li><a href="index.html">Home</a></li>
							<li>
								<a href="#" class="icon fa-angle-down">Eagle</a>
								<ul>
									<li><a href="quickstart.html">Quick Start</a></li>
									<li><a href="multithreadR.html">Multi-threaded R</a></li>
									<li><a href="instruction.html">Installation</a></li>
									<li><a href="videos.html">Videos</a></li>
									<li><a href="faq.html">faq</a></li>
<!--									<li>
										<a href="#">Submenu</a>
										<ul>
											<li><a href="#">Option One</a></li>
											<li><a href="#">Option Two</a></li>
											<li><a href="#">Option Three</a></li>
											<li><a href="#">Option Four</a></li>
										</ul> -->
									</li>
								</ul>
							</li>
							<li><a href="demo.html">Demo</a></li>
							<li><a href="contact.html">Contact Us</a></li>
							<!-- <li><a href="#" class="button">Contact Us</a></li> -->
						</ul>
					</nav>
				</header>

			
			
			<!-- Main -->
				<section id="main" class="container">
					<header>
						<h2>Frequently Asked Questions</h2>
						<p>Answers to the most commonly asked questions </p>
					</header>
		<!--			<div class="box">
						<span class="image featured"><img src="images/pic01.jpg" alt="" /></span>
						<h3>This is a subheading</h3>
						<p>Cep risus aliquam gravida cep ut lacus amet. Adipiscing faucibus nunc placerat. Tempus adipiscing turpis non blandit accumsan eget lacinia nunc integer interdum amet aliquam ut orci non col ut ut praesent. Semper amet interdum mi. Phasellus enim laoreet ac ac commodo faucibus faucibus. Curae ante vestibulum ante. Blandit. Ante accumsan nisi eu placerat gravida placerat adipiscing in risus fusce vitae ac mi accumsan nunc in accumsan tempor blandit aliquet aliquet lobortis. Ultricies blandit lobortis praesent turpis. Adipiscing accumsan adipiscing adipiscing ac lacinia cep. Orci blandit a iaculis adipiscing ac. Vivamus ornare laoreet odio vis praesent nunc lorem mi. Erat. Tempus sem faucibus ac id. Vis in blandit. Nascetur ultricies blandit ac. Arcu aliquam. Accumsan mi eget adipiscing nulla. Non vestibulum ac interdum condimentum semper commodo massa arcu.</p>
						<div class="row">
							<div class="6u 12u(mobilep)">
								<h3>And now a subheading</h3>
								<p>Adipiscing faucibus nunc placerat. Tempus adipiscing turpis non blandit accumsan eget lacinia nunc integer interdum amet aliquam ut orci non col ut ut praesent. Semper amet interdum mi. Phasellus enim laoreet ac ac commodo faucibus faucibus. Curae lorem ipsum adipiscing ac. Vivamus ornare laoreet odio vis praesent.</p>
							</div>
							<div class="6u 12u(mobilep)">
								<h3>And another subheading</h3>
								<p>Adipiscing faucibus nunc placerat. Tempus adipiscing turpis non blandit accumsan eget lacinia nunc integer interdum amet aliquam ut orci non col ut ut praesent. Semper amet interdum mi. Phasellus enim laoreet ac ac commodo faucibus faucibus. Curae lorem ipsum adipiscing ac. Vivamus ornare laoreet odio vis praesent.</p>
							</div>
						</div>
					</div>   -->
				
				
			  <div class="box">
			  <h3>General</h3>
					
			 
    <details>
    		<summary style="font-size: 20px" > How should I organize my input data? </summary>   		
    		
           <p style="font-size: 16px; color: grey" > Organize your data
           into three separate files; a file with the marker data, 
            a file with the phenotypic data, and a file with the marker map data. </p>
            <p style="font-size: 16px; color: grey" > The marker file has rows which 
            correspond to data on individuals and columns which 
            correspond to data on snp. Missing genotypes are allowed. The file does not have
            column headings. It is a space separated file. </p>
            <p style="font-size: 16px; color: grey" >
            The phenotype file has rows which correspond to data on individuals and columns which 
             correspond to data on traits  and fixed effects. There can be data on multiple traits
             in this file. Also, the file does not have to include fixed effects. It is optional 
             whether the first row of this file has the column headings. The file can contain 
             missing data. The file can be space or comma separated. </p>
            <p style="font-size: 16px; color: grey" >
            The marker map file has rows that correspond to snp. The number of rows in the 
            marker map file must be equal to the number of columns in the marker file. 
            It is optional if the first row of this file has column headings. This file cannot
            contain missing data. The file can be space or comma separated.  </p>
                       <p style="font-size: 16px; color: grey" >
             For further details, go to   <a href="quickstart.html">Quick Start guide</a>.
          </p>
     </details>
  
      
    <details>
    		<summary style="font-size: 20px" > Can Eagle deal with inbreds or can it only handle data on outbred individuals? </summary>
          
           <p style="font-size: 16px; color: grey"> Eagle can handle data recorded on inbred or outbred individuals but 
           we do assume all individuals are diploid.  </p>
     </details>
     
     
    <details>
    		<summary style="font-size: 20px" > What is the relationship between R and Eagle? </summary>
          
           <p style="font-size: 16px; color: grey" > 
           <a href="https://cran.r-project.org/">R</a> is the language upon which Eagle is
           built. However, much of the inner workings of Eagle are written in C++ and 
           interfaced with R via <a href="https://cran.r-project.org/web/packages/Rcpp/index.html">Rcpp</a> 
           and <a href="https://cran.r-project.org/web/packages/RcppEigen/index.html">RcppEigen</a>.
            </p>
     </details>

     
    <details>
    		<summary style="font-size: 20px" > Do I need a marker map? </summary>
           <p style="font-size: 16px; color: grey" > Eagle will run without a marker map being supplied. 
However, for interpretability of the results, it 
is best to supply a marker map. </p>
     </details>
     
    <details>
    		<summary style="font-size: 20px" > Will Eagle check for errors? </summary>
           <p style="font-size: 16px; color: grey" > 
           Eagle will check and report on some errors, especially with the input 
           files. However, there are errors that we cannot capture. For example, 
           we cannot check that the columns of snp data in the marker file are in the 
           same order as your snp in the marker map file. Also, we cannot check that the 
           row order of individuals in the phenotype file matches the row order of 
           individuals in the marker file. The best we can do is to check that the 
           column numbers and row numbers are consistent across the files. 
           
           </p>
     </details>


   <details>
    		<summary style="font-size: 20px" > Where can I go for help? </summary>
           <p style="font-size: 16px; color: grey" > 
         Your best resource for help is this website.  However, if you still need help, email us on 
         <a href="mailto:eaglehelp@csiro.au" target="_top">eaglehelp@csiro.au</a>
           
           </p>
     </details>


		
				</div>		      


<div class="box">
<h3>Using OpenGUI</h3>

   <details>
    		<summary style="font-size: 20px" >  I cannot see the file browser when I click on Choose File? </summary>
           <p style="font-size: 16px; color: grey" > This has been a source of frustration.  
           The file browser has opened, it is sitting in the background. 
This behaviour of the R file browser opening behind instead of in front of the active window 
is a known problem which has yet to be solved. </p>
     </details>

   <details>
    		<summary style="font-size: 20px" >  When I use OpenGUI, nothing happens </summary>
           <p style="font-size: 16px; color: grey" >Make sure you are running the latest version of R 
           along with updated versions of the packages. 
           Use the R function <code>update.packages(checkBuilt=TRUE, ask=FALSE)</code> to update your packages to the 
           latest version.
           </p>
     </details>

   <details>
    		<summary style="font-size: 20px" >  When I use OpenGUI, my web browser opens but
    		 everything is greyed out and I cannot click on any of the widgets </summary>
           <p style="font-size: 16px; color: grey" > Go back to the R window/console from 
           which you issued the <code>OpenGUI</code> command.  If an error has occurred, it will be 
           printed here.   The most likely cause is that your packages have been 
           installed under a version of R that is different to the one being used. 
            Run the following to update your installed packages 
            
<pre>
update.packages(checkBuilt=TRUE, ask=FALSE)
</pre>

</p>
     </details>
  
				
</div>


<div class="box">
<h3>Reading in Marker Genotypes </h3>


<details>
    		<summary style="font-size: 20px" > My marker data has heterozygous genotypes of the form 12 and 21. What do I do? </summary>
           <p style="font-size: 16px; color: grey" > 
           Unfortunately, Eagle can only handle a single value for heterozygous genotypes. This means that you will 
           need to replace one of the heterozygous genotypes with the other in your marker data set.  This might change in the future if 
           there is a demand for this functionality. </p>
           

</details>


   <details>
    		<summary style="font-size: 20px" >  What types of marker data can Eagle handle? </summary>
           <p style="font-size: 16px; color: grey" >Eagle can deal with two types of marker 
           data; genotypic and allelic. If marker genotypes are available, then they need to 
           be in 
a plain space separated text file where the rows are the individuals and the columns are the 
snps. The file should not contain row
or column headings. The genotypes can be any alphanumeric value, as long as these same alphanumeric 
values are used across all the loci. 
If allelic data are available, then this needs to be in <a href="http://www.cog-genomics.org/plink2/formats#ped">PLINK ped</a> 
form. Go to our <a href="quickstart.html">Quick Start guide</a> for further details. </p>
    </details>

   <details>
    		<summary style="font-size: 20px" >  Am I allowed to have missing marker data? </summary>
           <p style="font-size: 16px; color: grey" >Yes. If genotype data are available, 
           then any alphanumeric value can be used to denote missing data (as long
           as only a single value across all the loci is used). If allelic data are available, 
PLINK only allows 0 or - to be missing alleles. However, ideally, missing marker genotypes 
should be imputed with programs like 
<a href="https://faculty.washington.edu/browning/beagle/beagle.html">BEAGLE</a> or 
<a href="http://stephenslab.uchicago.edu/software.html">fastPHASE</a> before analysis. </p>
    </details>
    
   <details>
    		<summary style="font-size: 20px" > If I have missing marker genoytypes, what happens? </summary>
           <p style="font-size: 16px; color: grey" >Eagle sets the missing marker genotypes to the 
           heterozygous genotype AB. Since Eagle assumes an additive locus model, setting a missing genotype 
to AB is equivalent to imputing a genotype that has  no effect on a trait. 
If there are a large number of missing genotypes, this will reduce the power for detecting marker-trait associations. 
A better strategy is to  impute the missing marker genotypes prior to analysis with 
dedicated imputation software such as 
<a href="https://faculty.washington.edu/browning/beagle/beagle.html">BEAGLE</a> or 
<a href="http://stephenslab.uchicago.edu/software.html">fastPHASE</a>. </p>
    </details>


   <details>
    		<summary style="font-size: 20px" >  Can Eagle handle huge marker data sets? </summary>
           <p style="font-size: 16px; color: grey" >Yes. Eagle can analyse marker data 
           larger than the memory capacity of a computer by using out-of-memory matrix calculation. </p>
    </details>
    
    
   <details>
    		<summary style="font-size: 20px" >  I cannot get my marker data to read in </summary>
           <p style="font-size: 16px; color: grey" >
           A common source of error is an unequal number of elements across the rows of a marker file. 
           Eagle will capture this error, reporting on which row contains the unequal number of elements. </p>
           
           <p style="font-size: 16px; color: grey" >
           Check that there are no spaces at the beginning of a line.  </p> 
           
                                 <p style="font-size: 16px; color: grey" >
           Replace any tabs with spaces. </p>
           
           
            <p style="font-size: 16px; color: grey" >
            We've also encountered 
            problems when transferring files from a Windows system to a Unix system and vise versa.  
This is because the format for a Windows and a Unix text file differ slightly in how 
they handle the end of a line. When transferring 
files between different platforms, it is good practice to use a file conversion program 
first, such as <code>dos2unix</code> and <code>unix2dos</code>. </p>
    </details>

   <details>
    		<summary style="font-size: 20px" >  Can Eagle handle dominant marker loci and loci with more than two alleles? </summary>
           <p style="font-size: 16px; color: grey" >Yes. Suppose you have dominant marker data with genotype codes 0 and 1 for absence and presence, respectively. 
Then, when the marker data is being read with `Read Genotypes`, set the parameter AA to 0  and BB to 1 (or vise versa) but 
leave AB blank.   If you have a multi-allelic locus, say with 10 segregating alleles, then turn this locus into 10 
dominant loci and treat these loci as described above (remembering to modify the marker map file for the extra loci accordingly). </p>
    </details>
    
    


</div>




<div class="box">
<h3>Reading in Phenotypic Data  </h3>

 <details>
    		<summary style="font-size: 20px" > Can Eagle handle missing trait data and/or missing fixed effects data? </summary>
          <p style="font-size: 16px; color: grey" >Yes. Individuals with missing trait and/or fixed effects are removed from the analysis. 
However, only individuals whose data are being considered for analysis will be removed. Individuals with missing data not 
involved in the analysis will not be removed. 
   </details>



   <details>
    		<summary style="font-size: 20px" > Can my phenotypic file contain multiple traits? </summary>
          <p style="font-size: 16px; color: grey" >Yes. </p>
   </details>


   <details>
    		<summary style="font-size: 20px" >  Can my phenotypic file contain fixed effects that I may not use in an analysis? </summary>
          <p style="font-size: 16px; color: grey" >Yes </p>
   </details>


   <details>
    		<summary style="font-size: 20px" >  Do I need to be concerned with the order of 
    		the rows in my phenotypic file? </summary>
          <p style="font-size: 16px; color: grey" >Yes. The rows are ordered by 
          individual in the phenotypic file.  This same ordering of rows by individual 
          must be followed for the 
marker file. Eagle will check to make sure the total number of rows in the genotypic and 
phenotypic files match. However, 
Eagle  cannot check that the ordering of the rows in both files is the same.  </p>
   </details>



</div>



<div  class="box">
<h3>Running Eagle </h3>


   <details>
    		<summary style="font-size: 20px" >  Where can I get the latest version of Eagle? </summary>
    <p style="font-size: 16px; color: grey" >If Eagle has not been previously installed, then start  
    R and at the R prompt, type
<pre>
install.packages("Eagle", dependencies=TRUE)
</pre> </p>

<p style="font-size: 16px; color: grey">
Eagle is dependent upon several other packages. This command will install Eagle, along with any 
missing packages upon which Eagle is dependent. </p>

<p style="font-size: 16px; color: grey"> If Eagle is already installed, but you  want the latest version of Eagle and 
the packages upon which Eagle is dependent, then at 
the R prompt, type </p>
<pre>
update.packages(checkBuilt=TRUE, ask=FALSE) 
</pre>
</p>
  </details>


   <details>
    		<summary style="font-size: 20px" >  Do I need to update Eagle if I have updated my version of R? </summary>
    <p style="font-size: 16px; color: grey" > Yes. </p>

<p style="font-size: 16px; color: grey"> We have seen some strange behaviour, especially with `OpenGUI()`, when Eagle and its dependencies 
have been installed under different versions of R. When ever a newer version of R is installed, it is good practice to 
update it's packages with  </p>

<pre >
update.packages(checkBuilt=TRUE, ask=FALSE)
</pre>
</p>
  </details>


   <details>
    		<summary style="font-size: 20px" >  What platforms will Eagle run on? </summary>
    <p style="font-size: 16px; color: grey" > Eagle will run on the same platforms that R will 
    run on which are Linux, OS X (Mac), and Windows. </p>
  </details>




</div>




<div  class="box">
<h3>Performing Genome-wide Association Mapping  </h3>


   <details>
    		<summary style="font-size: 20px" >  Am I restricted to only analysing 
    		continuous traits or can I also analyse discrete/disease traits? </summary>
    <p style="font-size: 16px; color: grey" > 
    No, you are not restricted to only analysing continuous (quantitative) traits. 
    However, there may be a loss in 
    power for detecting marker-trait associations for a discrete trait.  
Eagle is based on linear mixed models. Linear mixed models assume a response (or trait) 
is normally distributed but they are 
robust to violations of their assumptions. </p>
  </details>
  
   <details>
    		<summary style="font-size: 20px" > Am I allowed to have interactions 
    		between my fixed effects? </summary>
    <p style="font-size: 16px; color: grey" > Yes, but using Eagle via OpenGUI can only 
    handle main effects. One solution is to add an 
extra column to the phenotypic data file that is the interaction of the two effects that 
are of interest. Then,  include these data 
as an extra fixed effect in the analysis. Alternately, from the R prompt, use the `AM` 
function and define the fixed effect part of the model 
by setting fformula. It would be best to look online for information on how to specify 
linear models in R which include interaction terms. </p>
  </details>
  
   <details>
    		<summary style="font-size: 20px" > My analysis doesn't seem to be  
    		using multiple threads/cpu </summary>
    <p style="font-size: 16px; color: grey" > Not all parts of an Eagle analysis is 
    parallelized. However, you should be seeing multiple threads being used at 
    different times
throughout each iteration of the model building process. 
It is most likely that your version of R is 
not making use of a multi-threaded BLAS library 
such as MKL or openBLAS. Look at our 
<a href="instruction.html">instruction notes</a> on how to install R that 
is  multi-threaded.  </p>
  </details>
  


</div>




		<!-- Scripts -->
			<script src="assets/js/jquery.min.js"></script>
			<script src="assets/js/jquery.dropotron.min.js"></script>
			<script src="assets/js/jquery.scrollgress.min.js"></script>
			<script src="assets/js/skel.min.js"></script>
			<script src="assets/js/util.js"></script>
			<!--[if lte IE 8]><script src="assets/js/ie/respond.min.js"></script><![endif]-->
			<script src="assets/js/main.js"></script>

	</body>
</html>