<!DOCTYPE HTML>
<!--
	Alpha by HTML5 UP
	html5up.net | @ajlkn
	Free for personal and commercial use under the CCA 3.0 license (html5up.net/license)
-->
<html>
	<head>
		<title>Eagle Quick Start</title>
		<meta charset="utf-8" />
		<meta name="viewport" content="width=device-width, initial-scale=1" />
		<!--[if lte IE 8]><script src="assets/js/ie/html5shiv.js"></script><![endif]-->
		<link rel="stylesheet" href="assets/css/main.css" />
		<!--[if lte IE 8]><link rel="stylesheet" href="assets/css/ie8.css" /><![endif]-->
	</head>
	<body>
		<div id="page-wrapper">

			<!-- Header -->
			
				<header id="header">
					<nav id="nav">
						<ul>
							<li><a href="index.html">Home</a></li>
							<li>
								<a href="#" class="icon fa-angle-down">Eagle</a>
								<ul>
									<li><a href="quickstart.html">Quick Start</a></li>
									<li><a href="multithreadR.html">Multi-threaded R</a></li>
									<li><a href="instruction.html">Installation</a></li>
									<li><a href="videos.html">Videos</a></li>
									<li><a href="faq.html">faq</a></li>
<!--									<li>
										<a href="#">Submenu</a>
										<ul>
											<li><a href="#">Option One</a></li>
											<li><a href="#">Option Two</a></li>
											<li><a href="#">Option Three</a></li>
											<li><a href="#">Option Four</a></li>
										</ul> -->
									</li>
								</ul>
							</li>
							<li><a href="demo.html">Demo</a></li>
							<li><a href="contact.html">Contact Us</a></li>
							<!-- <li><a href="#" class="button">Contact Us</a></li> -->
						</ul>
					</nav>
				</header>

			
			
			<!-- Main -->
				<section id="main" class="container">
					<header>
						<h2>Quick Start</h2>
						<p>A How To guide for using the Eagle GUI [revised May 2019] </p>
					</header>
		<!--			<div class="box">
						<span class="image featured"><img src="images/pic01.jpg" alt="" /></span>
						<h3>This is a subheading</h3>
						<p>Cep risus aliquam gravida cep ut lacus amet. Adipiscing faucibus nunc placerat. Tempus adipiscing turpis non blandit accumsan eget lacinia nunc integer interdum amet aliquam ut orci non col ut ut praesent. Semper amet interdum mi. Phasellus enim laoreet ac ac commodo faucibus faucibus. Curae ante vestibulum ante. Blandit. Ante accumsan nisi eu placerat gravida placerat adipiscing in risus fusce vitae ac mi accumsan nunc in accumsan tempor blandit aliquet aliquet lobortis. Ultricies blandit lobortis praesent turpis. Adipiscing accumsan adipiscing adipiscing ac lacinia cep. Orci blandit a iaculis adipiscing ac. Vivamus ornare laoreet odio vis praesent nunc lorem mi. Erat. Tempus sem faucibus ac id. Vis in blandit. Nascetur ultricies blandit ac. Arcu aliquam. Accumsan mi eget adipiscing nulla. Non vestibulum ac interdum condimentum semper commodo massa arcu.</p>
						<div class="row">
							<div class="6u 12u(mobilep)">
								<h3>And now a subheading</h3>
								<p>Adipiscing faucibus nunc placerat. Tempus adipiscing turpis non blandit accumsan eget lacinia nunc integer interdum amet aliquam ut orci non col ut ut praesent. Semper amet interdum mi. Phasellus enim laoreet ac ac commodo faucibus faucibus. Curae lorem ipsum adipiscing ac. Vivamus ornare laoreet odio vis praesent.</p>
							</div>
							<div class="6u 12u(mobilep)">
								<h3>And another subheading</h3>
								<p>Adipiscing faucibus nunc placerat. Tempus adipiscing turpis non blandit accumsan eget lacinia nunc integer interdum amet aliquam ut orci non col ut ut praesent. Semper amet interdum mi. Phasellus enim laoreet ac ac commodo faucibus faucibus. Curae lorem ipsum adipiscing ac. Vivamus ornare laoreet odio vis praesent.</p>
							</div>
						</div>
					</div>   -->
				<div class="box">
				
											
					  <h3>Getting Started</h3>
					  
					  <p> To install the Eagle package, from an interactive session in <a href="https://www.r-project.org/about.html">R</a>, type the following command   </p>
					    <pre class="r"><code> install.packages(&quot;data.table&quot;) </code></pre>
					    </p>
					    <p>
					 then type the command 
					   <pre class="r"><code> install.packages(&quot;Eagle&quot; , repos = &quot;https://cloud.r-project.org/&quot;, dependencies=TRUE) </code></pre>
 
 </p>
 <p>
 It may seem strange to first install data.table then Eagle. However, for R Open, under some linux distributions, 
 we found data.table failed to install when the package was obtained from https://cloud.r-project.org. The solution was to install data.table 
 from its default MRAN repository, and then install Eagle and its other dependencies from https://cloud.r-project.org. By using 
 this repository, we are downloading the most current version of Eagle (the version of MRAN is several versions and bugs old)
</p>

<p></p>
<p>
						    Once Eagle has been installed, the package is ready to be loaded into your R session with
                           <pre class="r"><code>library(&quot;Eagle&quot;)</code></pre>   
                           </p>
                           <p>
                           You will see a message saying that Eagle has been loaded. </p>								
						
				</div>	
					
				
				<div class="box">
				<h3>Running Eagle via the GUI</h3>
								
				<p> The easiest way of interacting with Eagle and performing an analysis is through 
					    our web-based graphical user interface (GUI). To run the GUI, at the R prompt, type
					    <pre class="r"><code>OpenGUI()</code></pre>
                        which will automatically open your web browser to our GUI. 
                </p>
                
				</div>	
					


				<div class="box">	
				<h3>The Eagle GUI home page - a good place to start</h3>
								
			    <p> When the Eagle GUI is run, it opens the web browser to a home page. 
				         At the top of the page is a navigation bar. To perform an analysis, 
				         work your way 
				         through each navigation item, from left to right, with the Read Z matrix item being only if you have 
				         repeated trait measurements for an individual and 
				        Read Map item being optional. </p>
				         
				
					    
			   <img src="images/home.png" width="80%">
                      
		       </div>				



               <div class="box">	
						
																
				<h3>How to organize your association mapping data</h3>
							
				
				
		    	<details>
         	   <summary style="font-size: 20px" > When an individual has a single trait value   </summary>
               <p style="font-size: 16px; color: grey" > 
             
             
             Suppose there are <em> n </em> individuals and for the trait of interest, each individual has a single observation, so 
             that there are <em> n </em> trait values.  Here, we split the data into three separate datasets:
             
											<ul>
											<li>a  <a href="#div1_markerdata">marker dataset</a>  </li>
											<li>a <a href="#div1_phenotypedata">phenotype dataset</a>  </li>
											<li>the <a href="#div1_markermapdata">marker map dataset</a>  </li>
			                                </ul>
		            
             
             
             
             
       
               <div class="box" id="div1_markerdata">	
								
			   <h3>Reading in the Marker Data</h3>
								
				<p> <font size="4" color="red">Note: From version 2.2, Eagle will also accept marker data in vcf format. 
				<a href="https://www.internationalgenome.org/wiki/Analysis/Variant%20Call%20Format/VCF%20%28Variant%20Call%20Format%29%20version%204.0/encoding-structural-variants">Variant Call Format version 4.0</a>
				is assumed. See our video on Input Data for more information.  </font>
				</p>
				
				
				
					    <p> There are two types of marker data file that Eagle can handle; 
					    a PLINK ped file and a plain text file. Below, is an example of each of these file types. </p>
					
                           <img src="images/ReadMarker_slide1.jpg" width="80%"  >
                    
                       <p> </p>
                       
  
                       
                       <p>  
                       To read in the marker data which are in PLINK format, click on the Read Genotype tab 
                       in the navigation bar at the top of the home page and complete the form (see below). After the upload file button has been clicked, a loading animation 
                        will appear in the bottom right of the page. Once the loading is complete, summary information will appear
                        on the right of the page.  </p>
 					   
                      <img src="images/ReadMarkerScreen_PLINK.jpg" width="80%">
                      
                       <p></p>
                       
                       <p> To read in the marker data which are in a plain text file, click on the Read Genotype tab 
                       in the navigation bar at the top of the home page and complete the form (see below). 
                       After the upload file button has been clicked, a loading animation 
                        will appear in the bottom right of the page. Once the loading is complete, summary information will appear
                        on the right of the page. 
                       </p>
                       
                       <p>
  					
                       <img src="images/ReadMarkerScreen_text.jpg" width="80%">
                    
                       </p>
                       
		       </div>			



               <div class="box"  id="div1_phenotypedata">	
																	
			   <h3>Reading in the Phenotypic Data</h3>
							
				    	<p> The file containing the phenotypic data is a plain text file. Below are 
				    	 three examples of what this file could contain. </p>
				    	
				    	<img src="images/ReadPheno_slide1.jpg" width="70%">
                       <p></p>
                       
                       <p> To read in the phenotypic data, click on the Read Phenotype tab in the navigation bar and 
                        complete the form (see below). After the upload file button has been clicked, a loading animation 
                        will appear in the bottom right of the page. Once the loading is complete, summary information will appear
                        on the right of the page. </p>

				    	<img src="images/ReadPhenoScreen.jpg" width="70%">

		       </div>			


               <div class="box"  id="div1_markermapdata" >	
						
																	
									<h3>Reading in the Marker Map</h3>
								
                        	<p> The file containing the marker map is a plain text file. Below are 
				    	 two examples of what this file could contain. </p>
				    	<img src="images/ReadMap_slide1.jpg" width="70%">
				    	
				    	
				    	
				    	<p> To read in the map data, click on the Read Map tab in the navigation bar and 
                        complete the form (see below). After the upload file button has been clicked, a loading animation 
                        will appear in the bottom right of the page. Once the loading is complete, summary information will appear
                        on the right of the page.  </p>
				    	<img src="images/ReadMapScreen.jpg" width="70%">
          
          
          
		       </div>		
		       
		       
	      
             
             
             
             
             
             
             
             
             
             
             
             
              </p>
              </details>
  		
				
				    	<details>
         	   <summary style="font-size: 20px" > When an individual has repeated trait values  </summary>
               <p style="font-size: 16px; color: grey" > 
               
               
               Suppose there are <em> m </em> individuals and for the trait of interest, there are 
               <em> n </em> trait values where  <em> n   > m </em>. Here, some individuals have repeated 
               observations for the trait. For example, we may have a plant study where the trait of interest is yield. 
               The study may contain <em> m </em> different plant varieties. For each variety, we may plant them in 
               multiple plots and record the yield per plot. 
               </p>
               
               <p>
               The data are split into four separate datasets:
               
											<ul>
											<li>a  <a href="#div2_markerdata">marker dataset</a>  </li>
											<li>a <a href="#div2_phenotypedata">phenotype dataset</a>  </li>
											<li>a <a href="#div2_markermapdata">marker map dataset</a>  </li>
											<li>a <a href="#div2_Zmatrix">Z matrix file</a>  </li>
											
			                                </ul>
		    
               
    
               <div class="box" id="div2_markerdata">	
								
			   <h3>Reading in the Marker Data</h3>
						
						<p> <font size="4" color="red">Note: From version 2.2, Eagle will also accept marker data in vcf format. 
				<a href="https://www.internationalgenome.org/wiki/Analysis/Variant%20Call%20Format/VCF%20%28Variant%20Call%20Format%29%20version%204.0/encoding-structural-variants">Variant Call Format version 4.0</a>
				is assumed. See our video on Input Data for more information.  </font>
				</p>		
				
					    <p> There are two types of marker data file that Eagle can handle; 
					    a PLINK ped file and a plain text file. Below, is an example of each of these file types. </p>
					
                           <img src="images/ReadMarker_slide1.jpg" width="80%"  >
                    
                       <p> </p>
                       
  
                       
                       <p>  
                       To read in the marker data which are in PLINK format, click on the Read Genotype tab 
                       in the navigation bar at the top of the home page and complete the form (see below). After the upload file button has been clicked, a loading animation 
                        will appear in the bottom right of the page. Once the loading is complete, summary information will appear
                        on the right of the page.  </p>
 					   
                      <img src="images/ReadMarkerScreen_PLINK.jpg" width="80%">
                      
                       <p></p>
                       
                       <p> To read in the marker data which are in a plain text file, click on the Read Genotype tab 
                       in the navigation bar at the top of the home page and complete the form (see below). After the upload file button has been clicked, a loading animation 
                        will appear in the bottom right of the page. Once the loading is complete, summary information will appear
                        on the right of the page. </p>
                       
                       <p>
  					
                       <img src="images/ReadMarkerScreen_text.jpg" width="80%">
                    
                       </p>
                       
		       </div>			



               <div class="box"  id="div2_phenotypedata">	
																	
			   <h3>Reading in the Phenotypic Data</h3>
							
				    	<p> The file containing the phenotypic data is a plain text file. Below are 
				    	 three examples of what this file could contain. </p>
				    	
				    	<img src="images/ReadPheno_slide1.jpg" width="70%">
                       <p></p>
                       
                       <p> To read in the phenotypic data, click on the Read Phenotype tab in the navigation bar and 
                        complete the form (see below). After the upload file button has been clicked, a loading animation 
                        will appear in the bottom right of the page. Once the loading is complete, summary information will appear
                        on the right of the page.  </p>

				    	<img src="images/ReadPhenoScreen.jpg" width="70%">

		       </div>			


               <div class="box"  id="div2_markermapdata" >	
						
																	
									<h3>Reading in the Marker Map</h3>
								
                        	<p> The file containing the marker map is a plain text file. Below are 
				    	 two examples of what this file could contain. </p>
				    	<img src="images/ReadMap_slide1.jpg" width="70%">
				    	
				    	
				    	
				    	<p> To read in the map data, click on the Read Map tab in the navigation bar and 
                        complete the form (see below). After the upload file button has been clicked, a loading animation 
                        will appear in the bottom right of the page. Once the loading is complete, summary information will appear
                        on the right of the page.  </p>
				    	<img src="images/ReadMapScreen.jpg" width="70%">
          
          
          
		       </div>		
		       
		         <div class="box"  id="div2_Zmatrix" >
	           
	           <h3>Reading in the Z matrix file</h3>
	           
	           <p> The Z matrix file is a plain text, space separate file with no row or column headings. Each row contains  
	           0's and a single entry that is 1. If the <em>i</em>th row and <em>j</em>th column  contains a 1, then this means the <em>i</em>th row of the phenotype file 
	           corresponds to the <em>j</em>th row of the marker file. A row can only contain a single entry with a value of 1. 
	           </p>
	           
	           <p>
	           To read the Z matrix file, click on the Z matrix tab 
	           in the navigation bar and 
                        complete the form (see below). After the upload file button has been clicked, a loading animation 
                        will appear in the bottom right of the page. Once the loading is complete, summary information will appear
                        on the right of the page.  </p>
				    	<img src="images/ReadZmatrixScreen.jpg" width="80%">
          
	      
	           </p>
	           
	           
	           
               </div>
               
               
              </p>
              </details>
  		
				
				
							
				
					  
				
		       </div>			

   	       
		       
		       	
               <div class="box"  id="div_analysis"  >	
							
																	
			   <h3>Performing Genome-wide Association Mapping</h3>
								
                    

                       <p>
                       We will assume that you have loaded your data successfully into Eagle and are now ready 
                       to analyse the data. Eagle finds SNP-trait associations by building the "best" model. 
                       We measure "best" via the extended BIC statistic. The false positive rate for this statistic 
                       is controlled by the lambda parameter. Values closer to 1 (0) reduce (increase) the false positive rate 
                       but also power of finding SNP-trait associations. 
                       
                    
                     
                       </p>
                       
                       
                       
                       <p>
                       Eagle gives you two ways of analysing your data. One 
                       way is where you set the lambda value manually. This is the "quick" way. The other way is where you 
                       let Eagle find the lambda value corresponding to a desired false positive rate for the analysis. This 
                       way is slower as it involves permutation but it is the preferred way. Both scenarios are presented below. 
                        </p>
                       
                       
                      
                       
                      <img style="float: left; padding: 0px 10px 10px 20px;" width="60%" src="images/Analysis_scenario1.jpg"/>
                      
                      <p>  <b>Scenario 1:</b>  Suppose you are interested in performing a "quick" preliminary analysis 
                      of your data. You can do this by selecting "Set manually" in Step 4.  
                      If the lambda value is left at its default value of 1, a  
                      conservative (low false positive rate) analysis of the data will be performed.  If you want to 
                      make the analysis less conservative, move the slider bar for the lambda value closer to 0.
                      This will increase the power of detecting SNP-trait associations but also increase 
                      the false positive rate. 
                      </p>
                    
                       
                       
                      <div style="clear: both"></div>
                       
                          <img style="float: left; padding: 0px 10px 10px 20px;" width="60%" src="images/Analysis_scenario2.jpg"/>
                      
                      <p>  <b>Scenario 2:</b>  Here, we have selected the "Set automatically" option in Step 4 
                      to let Eagle set the lambda value via permutation.  In this example, Eagle 
                      will find the lambda value that gives a false positive rate of 5% for finding SNP-trait associations. 
                      The analysis takes about five times as long to run.
                     
                      
                     
                      </p>
                      <div style="clear: both"></div>
                                        
                      <p></p>
                     <hr />

                      <h2> Findings </h2>
                      <p> Below are the two pages associated with Findings. The first page (A) 
                      is what appears by default.  It lists those snp, if any, that are 
                      in strongest linkage disequilibrium with the genes that are 
                      controlling the trait.  That is, the snp listed are the snp closest to 
                      the regions of the genome housing genes that influence the trait. 
                       </p>
                      
                      <p> The second page (B) gives additional information on the (additive) size of the snp effects, 
                      on their p-values, and an estimate of the amount of phenotypic variation explained by the snps. 
                      </p>
                      
    	           <img src="images/FindingsScreen.jpg" width="90%">

      
		
		       </div>			


	
				<div class="box"  id="div_1.5.0"  >	
	
  				<h2> Plots (Version 1.5.0)  </h2>
                
                
                
                <p> From Version 1.5.0, a Plot facility has been added to the GUI.
                At each step of the model building process, a set of score statistics are calculated. 
                These score statistics are used to identify the next "best" SNP in association with the trait.
                A user can examine how the score statistics change at each iteration of the iterative model building process. 
                </p>
                
                
                
                <p>
                An example screenshot is shown below. 
                </p>
                
                
            
    	           <img src="images/PlotScreen.png" width="90%">


	
	</div>
	
					
					
					
				</section>



			<!-- Footer -->
				<footer id="footer">
					<ul class="icons">
					    <li><a href="http://www.csiro.au"> <img src="images/csiro.jpg" 
										    class="img-responsive" width="100px" height="80px"/> </a> </li>
										    <li>  </li>
										    <li>  </li>
										    										    <li>  </li>
										    <li>  </li>
					<li><a href="http://data61.csiro.au"> <img src="images/data61.jpg" 
										     width="100px" height="80px"/> </a> </li>
										
				

					
					</ul>
					
				</footer>


		</div>

		<!-- Scripts -->
			<script src="assets/js/jquery.min.js"></script>
			<script src="assets/js/jquery.dropotron.min.js"></script>
			<script src="assets/js/jquery.scrollgress.min.js"></script>
			<script src="assets/js/skel.min.js"></script>
			<script src="assets/js/util.js"></script>
			<!--[if lte IE 8]><script src="assets/js/ie/respond.min.js"></script><![endif]-->
			<script src="assets/js/main.js"></script>

	</body>
</html>